### PR TITLE
Add migration for password algorithm change

### PR DESCRIPTION
**Pull request by zeripath**
#12688 changed the default for the user table leading to sync2 warnings

Unfortunately changing defaults requires a complete table rewrite in general.

However, just dropping columns could be bad - so this PR leverages the
techniques used in recreate table to recreate from the inferred schema
and recreates the user table.

This is not necessarily the correct thing to do - but code sometimes speaks
louder than words.

Signed-off-by: Andrew Thornton <art27@cantab.net>

### CODE REVIEW

File: models/migrations/v150.go [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+		_, err := x.Exec("ALTER TABLE `user` ALTER COLUMN passwd_hash_algo SET DEFAULT 'argon2';")
+		return err
+	case setting.Database.UseMSSQL:
+		// need to find the constraint and drop it, then recreate it.
+		sess := x.NewSession()
+		defer sess.Close()
+		if err := sess.Begin(); err != nil {
+			return err
+		}
+		res, err := sess.QueryString("SELECT [name] FROM sys.default_constraints WHERE parent_object_id=OBJECT_ID(?) AND COL_NAME(parent_object_id, parent_column_id)='?';", "user", "passwd_hash_algo")
```
Comment (zeripath) [inline models/migrations/v150.go]:
```suggestion
		res, err := sess.QueryString("SELECT [name] FROM sys.default_constraints WHERE parent_object_id=OBJECT_ID(?) AND COL_NAME(parent_object_id, parent_column_id)=?;", "user", "passwd_hash_algo")
```

File: models/migrations/v150.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+		return x.Sync2(new(User))
+	}
+	sess := x.NewSession()
+	defer sess.Close()
+	if err := sess.Begin(); err != nil {
+		return err
+	}
+
+	tempTableName := "tmp_recreate__user"
+	column.Default = "argon2"
```
Comment (zeripath) [inline models/migrations/v150.go]:
```suggestion
	column.Default = "'argon2'"
```

**Comment by silverwind** [comment]
Tested, worked for me.

**Comment by zeripath** [comment]
Need to double check what happens on sqlite - it may be that we drop all of the indexes and do not restore them.

**Comment by silverwind** [comment]
Maybe it's worth to move some of the code to (future) shared function?

**Comment by zeripath** [comment]
OK the indices are correct for the user table

**Comment by 6543** [comment]
@zeripath does **recreateTable** not cover this case?

**Comment by zeripath** [comment]
It's an option. I thought I'd write this to show what a non-recreate table solution looks like - this one means that any columns that are excess will not be deleted. Happy to drop the SQLite3 solution to use recreateTable though.

File: models/migrations/v150.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+		} else {
+			_, err := sess.Exec("ALTER TABLE [user] ADD DEFAULT('argon2') FOR passwd_hash_algo")
+			if err != nil {
+				return err
+			}
+		}
+		return sess.Commit()
+
+	case setting.Database.UseSQLite3:
+		// drop through
```
Comment (6543) [inline models/migrations/v150.go]: ok then lets use **recreateTable** for sqlite3 :)

Comment (6543) [reply models/migrations/v150.go]: as in discord: not jet :)

-> you can **resolve** this

**Review by 6543 — APPROVED** [review]

**Review by mrsdizzie — APPROVED** [review]

**Comment by 6543** [comment]
pleace resolve conflicts

File: models/migrations/v150.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+package migrations
+
+import (
+	"fmt"
+	"strings"
+
+	"code.gitea.io/gitea/modules/log"
+	"code.gitea.io/gitea/modules/setting"
+	"xorm.io/xorm"
```
Comment (lunny) [inline models/migrations/v150.go]: blank line.

**Review by lafriks — APPROVED** [review]

**Comment by 6543** [comment]
🚀

**Pull request merged**